app/controllers/subjects_controller.rb

```diff
@@ -56,6 +55,40 @@ def update
 
   private
 
+  def can_update_subject
+    if !check_for_removed_skills || (check_for_removed_skills && no_grades_using_skill)
+      if @subject.save
+        success(title: t(:subject_updated), text: t(:subject_updated_text, subject: @subject.subject_name))
+      else
+        return false
+      end
+    end
+
+    true
+  end
+
+  def no_grades_using_skill
+    destroyed_assignments = formatted_assignment_attributes.select { |v| v if v[:_destroy] == '1' }
+    lessons_using_subject = Lesson.where(subject_id: @subject.id).pluck(:id)
+
+    destroyed_assignments.each do |assignment|
+      if Grade.exists?(lesson_id: lessons_using_subject, skill_id: assignment[:skill_id])
+        failure(title: t(:unable_to_remove_skill_from_subject, skill_name: Skill.find(assignment[:skill_id]).skill_name), text: t(:skill_not_removed_because_grades))
+        return false
+      end
+    end
+
+    true
+  end
+
+  def check_for_removed_skills
```

Comment:
no need to do this or the `formatted_assignment_attributes` method.

After `@subject.assign_attributes subject_params` call in the `update` method you can just call `@subject.assignments.filter(&:marked_for_destruction?)` will return an array of assignments that were marked for destruction

Example from the debugger after I removed one skill from the subject: 

```ruby
@subject.assignments.filter(&:marked_for_destruction?)
# => [#<Assignment id: 100, skill_id: 1, subject_id: 15, deleted_at: nil, created_at: "2019-01-05 16:06:08.976944000 +0000", updated_at: "2019-01-05 16:06:08.976944000 +0000">]
```